src/kernels/MIOpenInterpolate.cpp

```diff
@@ -0,0 +1,982 @@
+            {
+                FLOAT_ACCUM w_lambda =
+                    compute_back_lambda(k, w, scale_factor_w_, Win, Wout, align_corners);
+                tensor_layout_t<5> output_layout(n, c, i, j, k);
+
+                output += CVT_FLOAT2ACCUM(
+                              output_grad[output_grad_tv.get_tensor_view_idx(output_layout)]) *
+                          d_lambda * h_lambda * w_lambda;
+            }
+        }
+    }
+    input_grad[input_grad_tv.get_tensor_view_idx(tensor_layout)] = CVT_ACCUM2FLOAT(output);
+}
+
+extern "C" __global__ void InterpolateTrilinearBackward(OUTPUT_TYPE* __restrict__ input_grad,
+                                                        const INPUT_TYPE* __restrict__ output_grad,
+                                                        const tensor_view_t<5> input_grad_tv,
+                                                        const tensor_view_t<5> output_grad_tv,
+                                                        const size_t nelems,
+                                                        const float* scale_factors,
+                                                        const bool align_corners)
+{
+    interpolateTrilinearBackward<INPUT_TYPE, OUTPUT_TYPE>(input_grad,
+                                                          output_grad,
+                                                          input_grad_tv,
+                                                          output_grad_tv,
+                                                          nelems,
+                                                          scale_factors,
+                                                          align_corners);
+}
+
+__device__ inline FLOAT_ACCUM
+compute_scales_value(FLOAT_ACCUM scale, int64_t input_size, int64_t output_size)
+{
+    return (scale == 0.f) ? (static_cast<FLOAT_ACCUM>(input_size) / output_size) : (1.0f / scale);
+}
+
+__device__ inline int64_t
+nearest_idx(int64_t output_index, int64_t input_size, int64_t output_size, FLOAT_ACCUM scales)
+{
+    if(output_size == input_size)
+    {
+        return output_index;
+    }
+    else if(output_size == 2 * input_size)
+    {
+        return output_index / 2;
+    }
+    else
+    {
+        FLOAT_ACCUM scale = compute_scales_value(scales, input_size, output_size);
+        return min(static_cast<int64_t>((output_index * scale)), input_size);
+    }
+}
+
+template <typename TI, typename TO>
+__device__ inline void interpolateNearestForward(const TI* __restrict__ input,
+                                                 TO* __restrict__ output,
+                                                 const tensor_view_t<5> input_tv,
+                                                 const tensor_view_t<5> output_tv,
+                                                 const size_t nelems,
+                                                 const float* scale_factors)
+{
+    int64_t gid = blockIdx.x * blockDim.x + threadIdx.x;
```

Comment:
the original kernel in modnn use long, and the type in cpu code I use is also long, so I change it to int64_t to avoid wrong computing (which did happen)